docs/design/secure-config.md

```diff
@@ -0,0 +1,228 @@
+      config.toml
+  workspaces/
+    def456/
+      config.toml
+      metadata.binpb
+my-repo/.jj/
+  workspace-id (contains "def456")
+  workspace-config.toml (symlink to $HOME/.config/jj/workspaces/def456/config.toml)
+  repo/
+    repo-id (contains "abc123")
+    config.toml (symlink to $HOME/.config/jj/repos/abc123/config.toml)
+```
+Metadata.binpb will refer to the following protobuf:
+
+```proto
+message Metadata {
+  // This is used to distinguish between copies and moves.
+  string path = 1;
+}
+```
+
+The function to load repository configuration, will roughly speaking, look like:
+```rust
+enum ConfigLoadError {
+    NoRepoId,
+    NoConfig,
+    PathMismatch,
+}
+
+fn load_repo_config_path(repo: Path) -> Result<PathBuf, ConfigLoadError> {
+    let Ok(repo_id) = std::fs::read_to_string(repo.join("repo-id")) else return Err(NoRepoId);
+    let repo_config_dir = config_dir.join("repos").join(repo_id);
+    let Ok(metadaata) = Metadata::decode(std::fs::read(repo_config_dir.join("metadata.binpb"))) else return Err(NoConfig);
+    if metadata.path != repo {
+        return Err(PathMismatch)
+    }
+    Ok(repo_config_dir.join("config.toml"))
+}
+```
+
+#### No repo-ID file
+To preserve backwards compatibility, we will introduce a period of auto-migration. The current plan is 12 jj versions (approximately 1 year). During this period, if a repo ID has not yet been generated, we will silently perform the following (order matters, to ensure failure halfway through doesn’t affect things):
+
+##### With config files
+1. Generate a repo ID `abc123`
+2. Create `$HOME/.config/jj/repo/abc123/metadata.binpb`
+3. Create `$HOME/.config/jj/repo/abc123/config.toml` as a copy of the original config file
+4. Atomically generate a Repo-ID file containing `abc123`
+5. Remove the original config file
+6. Symlink the original config file to the new config file (on windows, this might fail, so if it fails we'll create a readonly copy of the config with the header `# Do not edit. Run jj config edit --repo to edit your config, or edit $PATH`)
+7. Print a warning that the old config file existed on disk, that it has been moved, and to run `jj config edit --repo` in the future instead of modifying the file directly.
+
+After the migration period is over, we will stop the auto-migration, and only print the warning.
+
+##### Without config files
+If there is no config files, then we don't care.
+
+#### No config directory for the repo-ID
+This could occur, for example, if the user created a repository in linux, rebooted into windows on the same computer, and attempted to access that repo.
+
+In this event, the user probably expects their config to be attached to the repository, and they expect it to still work on linux, so we will:
+* Create a new `$HOME/.config/jj/repo/$REPO_ID` directory for the same repo ID (to ensure that the config still works on windows)
+* Print a warning that if there was any per-repo configuration, it is no longer available.
```

Comment:
Same here: This seems like a special case. Better to explain the normal case first. Oh, I just realized that you actually did explain the normal case above but only in a code snippet. I skipped over that because if was looking for a textual description. Can we add that?